packages/structures/src/bitfields/BitField.ts

```diff
@@ -0,0 +1,203 @@
+import type { EnumLike, NonAbstract, RecursiveReadonlyArray } from '../utils/types';
+
+/**
+ * Data that can be resolved to give a bitfield. This can be:
+ * A bit number (this can be a number literal or a value taken from {@link (BitField:class).Flags})
+ * A string bit number
+ * An instance of BitField
+ * An Array of BitFieldResolvable
+ */
+export type BitFieldResolvable<Flags extends string> =
+	| Flags
+	| Readonly<BitField<Flags>>
+	| RecursiveReadonlyArray<Flags | Readonly<BitField<Flags>> | bigint | number | `${bigint}`>
+	| bigint
+	| number
+	| `${bigint}`;
+
+/**
+ * Data structure that makes it easy to interact with a bitfield.
+ */
+export abstract class BitField<Flags extends string> {
+	/**
+	 * Numeric bitfield flags.
+	 * <info>Defined in extension classes</info>
+	 */
+	public static readonly Flags: EnumLike<unknown, bigint | number> = {};
+
+	public static readonly DefaultBit: bigint = 0n;
+
+	/**
+	 * Bitfield of the packed bits
+	 */
+	public bitfield: bigint;
+
+	declare public ['constructor']: NonAbstract<typeof BitField<Flags>>;
```

Comment:
Do you need the NonAbstract type at all?

Reply:
Yes, else `new this.constructor(...)` errors because can't instantiate abstract classes.